### PR TITLE
Exporting CSV Delta File state resource to accept array of functions

### PR DESCRIPTION
**Pull request by meenahoda**
Story details: https://app.clubhouse.io/wmfs/story/405

### CODE REVIEW

File: lib/Transformer.js [hunk 1]
```diff
@@ -62,15 +63,34 @@ class Transformer extends Transform {
     const outputValues = this.transformers
       .map(fn => fn(sourceRow, this.info))
 
-    if (!this.filterFn(outputValues)) {
+    const filterFns = Array.isArray(this.filterFn) ? this.filterFn : [this.filterFn]
```
Comment (jezhiggins) [inline lib/Transformer.js]: Can we hoist this up to line 13?

```
const filters = options.filterFunction ? options.filterFunction : () => true
this.filterFn = Array.isArray(this.filterFn) ? this.filterFn : [this.filterFn]
```

File: lib/Transformer.js [hunk 2]
```diff
@@ -62,15 +63,34 @@ class Transformer extends Transform {
...
+
+    let filter = true
+
+    filterFns.forEach(fn => {
+      if (!fn(outputValues)) filter = false
+    })
+
+    if (!filter) {
```
Comment (jezhiggins) [inline lib/Transformer.js]: We can collapse this down using Array.every 
```
if (!filterFns.every(f => f(outputValues))) {
  this.filterRowCount()
  return callback(null, null)
}
```

Comment (meenahoda) [reply lib/Transformer.js]: I was trying to find something like this! Thanks

File: lib/Transformer.js (hunk 1 above)
Comment (meenahoda) [reply lib/Transformer.js]: Will do

File: lib/Transformer.js [hunk 3]
```diff
@@ -62,15 +63,34 @@ class Transformer extends Transform {
...
       this.filterRowCount()
       return callback(null, null)
     }
 
-    this.transformerFn(
-      outputValues,
-      (err, transformedValues) => callback(err, csvEncoder(transformedValues))
-    )
+    const transformFns = Array.isArray(this.transformerFn) ? this.transformerFn : [this.transformerFn]
```
Comment (jezhiggins) [inline lib/Transformer.js]: Again, I'd bump this up to the initialisation phase of things.

File: lib/Transformer.js [hunk 4]
```diff
@@ -62,15 +63,34 @@ class Transformer extends Transform {
...
+    transformFns.forEach(fn => {
+      return new Promise((resolve, reject) => {
+        fn(transformedValues, (err, values) => {
+          if (err) return reject(err)
+          transformedValues = values
+          resolve()
+        })
+      })
+    })
+
```
Comment (jezhiggins) [inline lib/Transformer.js]: I am pretty confident that if this is working, then it's working by accident. I am not sure what the correct solution is, but let me have think.

Comment (jezhiggins) [reply lib/Transformer.js]: We need to pass the values through each transformer in turn, and once we've done all that ping it back out through `callback`. Our transfomer function interface isn't great, but we can work with it. Something like this should do it, I think
```
const transformerCascade = (err, values, index = 0) => {
  if (index !== transformFns.length && !err) {
     transformFns[index](values, (e, transformedValues) => transformerCascade(e, transformedValues, index+1))
  } else {
    callback(err, csvEncoder(values))
   }
 }

transformerCascade(null, outputValues)
```

Comment (jezhiggins) [reply lib/Transformer.js]: Actually we could wrap that up a bit and push it up the initialisation section too - 
```
this.transformers = (outputValues, cb) => {
  const transformerCascade = (err, values, index = 0) => {
    if (index !== transformFns.length && !err) {
       transformFns[index](values, (e, transformedValues) => transformerCascade(e, transformedValues, index+1))
    } else {
      cb(err, values)
    }
  }

  transformerCascade(null, outputValues)
}
```
then the call inside `_transform` collapses to 
```
this.transformers(
  outputValues,
  (cb, transformedValues) => callback(err, csvEncoder(transformedValues))
)
```

Comment (meenahoda) [reply lib/Transformer.js]: Thank you!

Comment (meenahoda) [reply lib/Transformer.js]: Thank you!

**Pull request merged**